Error handling

rcky opened the pull request:
Mainly added the error handling and the service name.

rcky committed:
fix: add url normalization

rcky committed:
chore: adds editorconfig

rcky committed:
feat(microservice-name): the service now takes a service name out of its boot options

rcky committed:
feat(error-handler): adds wrapper for custom error handler

rcky committed:
feat(service-name): adds service name to the microservice

rcky committed:
style: cleanup error messages with unwanted indentation

rcky committed:
style: short linting

lkappeler commented on src/MicroserviceErrorHandler.js:
```diff
@@ -0,0 +1,44 @@
+        if(!serviceName) {
+            const msg = 'Configuration value for @joinbox/loopback-microservice#errorHandler' +
+                ' is missing: Please add the property "serviceName" to the "params" section of the' +
+                ' middleware config or copy your general microservice config (containing the name)'
```
you forgot the plus here

rcky replied:
Ugh!

lkappeler commented on src/MicroserviceApiClient.js:
```diff
@@ -4,7 +4,7 @@ const url = require('url');
 
     constructor(baseUrl, {requestLibrary = superagent} = {}) {
-        this.base = baseUrl;
+        this.base = this._normalizeBase(baseUrl);
```
Is there a reason to flag this function as internal in think it could be handy in the tests in some cases.

rcky replied:
Nope, i just didn't see any use case. But we can easily _make it public_

lkappeler commented on test/integration/Microservice.test.js:
```diff
@@ -43,18 +42,36 @@ describe('The Microservice', function() {
+                .api
+                .get('/nonExisting')
+                .set('accept', 'application/json');
+            return Promise.reject(new Error('Request to non existing route should fail.'))
```
Why do you need to reject the promise manually here? Shouldn't the request reject it?

rcky replied:
Yeah, it should, but if it does not, the test will not fail (and if i throw it, it would be caught by the catch clause and give me an error message which does not tell me why the test failed).

lkappeler commented on test/unit/MicroserviceErrorHandler.test.js:
```diff
@@ -0,0 +1,90 @@
+        expect(handler).to.have.property('serviceName', serviceName);
+    });
+
+    it('can be instantiated with options containing the microservice config object', function(){
```
What is the reason for the two interfaces to configure the name?

rcky replied:
Basically you could use the error handler without having our Microservice involved and then you probably don't want to add `{ "microservice": { "name": "blubb" }}` to your config.

lkappeler commented on test/unit/MicroserviceErrorHandler.test.js:
```diff
@@ -0,0 +1,90 @@
+
+    it('throws and error if it cannot extract a service name (there is no reason to use it otherwise' +
+        ' at this time', function(){
+        expect(() => new MicroserviceErrorHandler({})).to.throw();
```
May you should check what the ErrorHandler throws.

rcky replied:
Basically I don't care ;) I just want the application to not be able to boot.

lkappeler reviewed (commented):
The code looks good to me in general. Found some small issues may you want to have a look.

rcky committed:
fix(review): fix code according to review

rcky commented:
@lkappeler updated

lkappeler commented on src/MicroserviceErrorHandler.js:
```diff
@@ -37,8 +39,8 @@ module.exports = class MicroserviceErrorHandler {
     createErrorHandler() {
         return (err, req, res, next) => {
-            err.serviceName = this.serviceName;
+            err.serviceName = this.serviceName; // eslint-disable-line no-param-reassign
```
I prefer the ` eslint-disable-next-line`.

lkappeler commented on src/MicroserviceErrorHandler.js:
```diff
@@ -15,19 +15,21 @@ module.exports = class MicroserviceErrorHandler {
     }
 
-    _copyAndAddSafeFields(options, fields) {
+    _copyAndAddSafeFields(options, fields = []) {
```
Whats the intention of copying the options?

rcky replied:
We get this options by reference (usually via configuration). I don't know where they come from and where they are used outside of the error handler. Modifying them by reference might lead to side-effects (especially if we plan to extend these options in the future).

rcky replied:
Btw: I just saw that i still modify them wrongly (below when i push into the array), will fix that.

lkappeler reviewed (approved):
The changes are ok for me. Commented some small things.

rcky committed:
fix: adds proper copying of passed configuration values

rcky committed:
feat(boot-options): changes the options interface of the Microservice.

The microservice constructor now takes an options object. While this object previously contained the boot options for Loopback, these configuration is now moved to the boot property of the options object.

BREAKING CHANGE: Directly passing the boot options to the constructor and the helper methods to boot and start the service will no longer work.

lkappeler commented:
@rcky Im op with the last changes you did.
